toolkit/tools/pkg/imagecustomizerlib/liveosisobuilder.go

```diff
@@ -243,6 +246,45 @@ func (b *LiveOSIsoBuilder) prepareRootfsForDracut(writeableRootfsDir string) err
 	return nil
 }
 
+// canonalizeExtraCommandLine
+//
+// given the kernel extra commandline parameters, this function ensures
+// that they are well-formatted for grub correct consumption.
+//
+// inputs:
+//   - extraCommandLine:
+//     extra commandline parameters to canonalize.
+//
+// outputs
+//   - canonalized commadline parameters.
+func canonalizeExtraCommandLine(extraCommandLine string) string {
+	cannonicalExtraCommandLine := ""
+	// Split by spaces/tabs into 'fields'.
+	fields := strings.Fields(extraCommandLine)
+	for _, field := range fields {
+		// If a field starts with 'ds=nocloud;s=' (for cloud-init), it means
+		// the user did not wrap it in single quotes as it should be. If left
+		// as-is, the grub parsing at runtime will drop the text after the ';'.
+		//
+		// Unfortunately, if the user adds single quotes, the higher-level
+		// toolkit and MIC schema validation fail because the Mariner Toolkit
+		// does not allow it.
+		//
+		// This Toolkit limitation is because it uses 'sed' to manipulate
+		// grub.cfg and 'sed' relies on single quotes to parse its input
+		// expression. So, include a single quote will break the sed input
+		// parsing.
+		//
+		// The work around below is to restore the needed single quotes.
+		if strings.HasPrefix(field, "ds=nocloud;s=") {
```

Comment:
MIC is going to support strong back-compat. So, hacks like this aren't permitted.

The BASH/grub rules for parsing single quotes are actually very straightforward since there aren't any escape characters inside single-quote strings. So, it shouldn't be too difficult to expand the validity checking to correctly handle single-quotes.

Reply:
for some reason, i misread the sed character to be the single quote instead of the back-tick here:
https://github.com/microsoft/CBL-Mariner/blob/14f37351044fc9b943c3b9660957f3f97e4530c8/toolkit/tools/imagecustomizerapi/kernelcommandline.go#L33
Will revise...